File: app/android/app/src/main/java/com/ambientnotes/app/recognition/providers/AudDProvider.kt
package com.ambientnotes.app.recognition.providers

import com.ambientnotes.app.recognition.RecognitionProvider
import com.ambientnotes.app.recognition.RecognitionProviderException
import com.ambientnotes.app.recognition.RecognitionResult
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.MultipartBody
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject

/**
 * Talks to the user's own AudD (https://audd.io) account using their API
 * token, entered in Settings. BYO-account, same as ACRCloud.
 */
class AudDProvider(
    private val apiTokenProvider: suspend () -> String?,
    private val httpClient: OkHttpClient = OkHttpClient(),
) : RecognitionProvider {

    override val id: String = "audd"
    override val displayName: String = "AudD"

    override suspend fun isConfigured(): Boolean = !apiTokenProvider().isNullOrBlank()

    override suspend fun recognize(pcm16Audio: ByteArray, sampleRateHz: Int): RecognitionResult =
        withContext(Dispatchers.IO) {
            val token = apiTokenProvider().takeUnless { it.isNullOrBlank() }
                ?: return@withContext RecognitionResult.noMatch(id)

            val body = MultipartBody.Builder()
                .setType(MultipartBody.FORM)
                .addFormDataPart("api_token", token)
                .addFormDataPart("return", "apple_music,spotify")
                .addFormDataPart(
                    "file", "sample.wav",
                    pcm16Audio.toRequestBody("audio/wav".toMediaType())
                )
                .build()

            val request = Request.Builder().url("https://api.audd.io/").post(body).build()

            val response = try {
                httpClient.newCall(request).execute()
            } catch (e: java.io.IOException) {
                throw RecognitionProviderException(id, "Network error contacting AudD", e)
            }

            response.use {
                if (!it.isSuccessful) throw RecognitionProviderException(id, "AudD HTTP ${it.code}")
                parseResponse(JSONObject(it.body?.string().orEmpty()))
            }
        }

    internal fun parseResponse(json: JSONObject): RecognitionResult {
        if (json.optString("status") != "success" || json.isNull("result")) {
            return RecognitionResult.noMatch(id)
        }
        val result = json.optJSONObject("result") ?: return RecognitionResult.noMatch(id)

        val externalIds = mutableMapOf<String, String>()
        result.optJSONObject("spotify")?.optString("id")?.let { if (it.isNotBlank()) externalIds["spotify"] = it }
        result.optJSONObject("apple_music")?.optString("url")?.let { if (it.isNotBlank()) externalIds["apple_music"] = it }

        return RecognitionResult(
            matched = true,
            title = result.optString("title").takeIf { it.isNotBlank() },
            artist = result.optString("artist").takeIf { it.isNotBlank() },
            album = result.optString("album").takeIf { it.isNotBlank() },
            releaseDate = result.optString("release_date").takeIf { it.isNotBlank() },
            confidence = 0.9f, // AudD does not return a numeric confidence score
            providerName = id,
            externalIds = externalIds,
        )
    }
}

File: service/app/providers/audd_provider.py
from __future__ import annotations

import logging

import httpx

from app.config import Settings
from app.models import RecognitionResult
from app.providers.base import RecognitionProvider
from app.providers.registry import register_provider

logger = logging.getLogger(__name__)

AUDD_ENDPOINT = "https://api.audd.io/"


@register_provider("audd")
class AudDProvider(RecognitionProvider):
    def __init__(self, api_token: str, enabled: bool = True):
        super().__init__()
        self.api_token = api_token
        self.enabled = enabled

    @classmethod
    def from_settings(cls, settings: Settings) -> "AudDProvider":
        return cls(api_token=settings.audd_api_token, enabled=settings.audd_enabled)

    async def is_available(self) -> bool:
        return bool(self.enabled and self.api_token)

    async def recognize(self, audio_bytes: bytes, sample_rate_hz: int) -> RecognitionResult:
        if not await self.is_available():
            return RecognitionResult(matched=False, provider=self.name)

        files = {"file": ("sample.wav", audio_bytes, "audio/wav")}
        data = {"api_token": self.api_token, "return": "apple_music,spotify"}

        async with httpx.AsyncClient(timeout=15) as client:
            resp = await client.post(AUDD_ENDPOINT, data=data, files=files)
            resp.raise_for_status()
            payload = resp.json()

        if payload.get("status") != "success" or not payload.get("result"):
            return RecognitionResult(matched=False, provider=self.name, raw_provider_response=payload)

        result = payload["result"]
        external_ids = {}
        if spotify := result.get("spotify"):
            external_ids["spotify"] = spotify.get("id", "")
        if apple := result.get("apple_music"):
            external_ids["apple_music"] = apple.get("url", "")

        return RecognitionResult(
            matched=True,
            title=result.get("title"),
            artist=result.get("artist"),
            album=result.get("album"),
            release_date=result.get("release_date"),
            confidence=0.9,  # AudD does not return a numeric confidence score
            provider=self.name,
            external_ids=external_ids,
            raw_provider_response=payload,
        )
